presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+                    newFileLength += remainingFollowingFileBytes.length;
+                }
+            }
+        }
+        catch (IOException e) {
+            log.warn(e, "%s encountered an error while flushing file %s", Thread.currentThread().getName(), newFilePath);
+            tryDeleteFile(newFilePath);
+            return false;
+        }
+
+        boolean updated;
+        Set<Path> cacheFilesToDelete = new HashSet<>();
+        Lock writeLock = persisted.get(targetFile).getLock().writeLock();
+        writeLock.lock();
+        try {
+            RangeMap<Long, LocalCacheFile> cache = persisted.get(targetFile).getRange();
+            // check again if the previous or following range has been updated by someone else
+            LocalCacheFile newPreviousCacheFile = cache.get(key.getOffset() - 1);
+            LocalCacheFile newFollowingCacheFile = cache.get(key.getOffset() + key.getLength());
+
+            if (!cacheFileEquals(previousCacheFile, newPreviousCacheFile) || !cacheFileEquals(followingCacheFile, newFollowingCacheFile)) {
+                // someone else has updated the cache; delete the newly created file
+                updated = false;
+            }
+            else {
+                updated = true;
+
+                // remove all the files that can be covered by the current range
+                cacheFilesToDelete = cache.subRangeMap(Range.closedOpen(key.getOffset(), key.getOffset() + key.getLength())).asMapOfRanges().values().stream()
+                        .map(LocalCacheFile::getPath).collect(Collectors.toSet());
+
+                // update the range
+                Range<Long> newRange = Range.closedOpen(newFileOffset, newFileOffset + newFileLength);
+                cache.remove(newRange);
+                cache.put(newRange, new LocalCacheFile(newFileOffset, newFilePath));
+            }
+        }
+        finally {
+            writeLock.unlock();
+        }
+
+        // no lock is needed for the following operation
+        if (updated) {
+            // remove the the previous or following file as well
+            if (previousCacheFile != null) {
+                cacheFilesToDelete.add(previousCacheFile.getPath());
+            }
+            if (followingCacheFile != null) {
+                cacheFilesToDelete.add(followingCacheFile.getPath());
+            }
+        }
+        else {
+            cacheFilesToDelete = ImmutableSet.of(newFilePath);
+        }
+
+        cacheFilesToDelete.forEach(LocalRangeCacheManager::tryDeleteFile);
+        return true;
+    }
+
+    private static void tryDeleteFile(Path path)
+    {
+        try {
+            File file = new File(path.toUri());
+            if (file.exists()) {
+                Files.delete(file.toPath());
+            }
+        }
+        catch (IOException e) {
+            // ignore
+        }
+    }
+
+    private static boolean cacheFileEquals(LocalCacheFile left, LocalCacheFile right)
```

Comment:
Curious: Can we just use `equals`?

Reply:
We need to handle null cases.

Reply:
The null case seems to be handled by `equals` as well =/

```
            if (this == o) {  // both null
                return true;
            }
            if (o == null || getClass() != o.getClass()) {  // either one is null
                return false;
            }
```

Reply:
Actually, let me check if `null == null` is true 😂

Reply:
What if `left` is null?